File: project_ws/src/robot_exploration/robot_exploration/service/client_service_test_node.py
import rclpy
from rclpy.node import Node
from std_srvs.srv import SetBool

class IdentifyClient(Node):

    def __init__(self):
        super().__init__('identify_client')
        self.cli = self.create_client(SetBool, 'identify_robot')
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Service not available, waiting...')
        self.req = SetBool.Request()

    def send_request(self, identify: bool):
        self.req.data = identify
        return self.cli.call_async(self.req)

def main(args=None):
    rclpy.init(args=args)
    node = IdentifyClient()
    future = node.send_request(True)  # On envoie toujours "identifier"
    rclpy.spin_until_future_complete(node, future)
    if future.result() is not None:
        node.get_logger().info(
            f"Réponse: success={future.result().success}, message='{future.result().message}'"
        )
    else:
        node.get_logger().error("Service call failed ")
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: project_ws/src/robot_exploration/robot_exploration/service/client_service_test_node.js
#!/usr/bin/env node
'use strict';

// Node.js client equivalent of the Python rclpy client.
// Calls std_srvs/srv/Trigger service named 'identify_robot'.

const rclnodejs = require('rclnodejs');

async function main() {
  await rclnodejs.init();

  const node = rclnodejs.createNode('identify_client');
  const Trigger = rclnodejs.require('std_srvs').srv.Trigger;

  const client = node.createClient(Trigger, 'identify_robot');

  rclnodejs.spin(node);

  // Wait for the service to be available (polling every 1s)
  // client.waitForService(timeout) resolves to true if available within timeout
  // Fallback to retry loop if it returns false.
  // Some older versions may throw; catch to keep looping.
  // eslint-disable-next-line no-constant-condition
  while (true) {
    try {
      const ready = await client.waitForService(1000);
      if (ready) break;
    } catch (_) {
      // ignore and keep trying
    }
    node.getLogger().info('Service not available, waiting...');
  }

  const request = new Trigger.Request();

  // Use callback style for broad compatibility across rclnodejs versions
  await new Promise((resolve) => {
    client.sendRequest(request, (response) => {
      if (response) {
        node.getLogger().info(
          `Réponse: success=${response.success}, message='${response.message}'`
        );
      } else {
        node.getLogger().error('Service call failed');
      }
      resolve();
    });
  });

  // Give a brief moment for logs to flush
  setTimeout(() => {
    node.destroy();
    rclnodejs.shutdown();
  }, 10);
}

main().catch((err) => {
  // Last-resort error handler
  // eslint-disable-next-line no-console
  console.error(err);
  process.exit(1);
});

